common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -128,49 +151,100 @@ protected AppShuffleInfo validateAndGetAppShuffleInfo(String appId) {
+          // current incoming one
+          throw new RuntimeException(String.format("Rejecting shuffle blocks push request for"
+            + " shuffle %s with shuffleMergeId %s for application %s_%s as a higher"
+              + " shuffleMergeId %s request is already seen", shuffleId, shuffleMergeId,
+              appShuffleInfo.appId, appShuffleInfo.attemptId, latestShuffleMergeId));
         } else {
-          return map;
+          // Higher shuffleMergeId seen for the shuffle ID meaning new stage attempt is being
+          // run for the shuffle ID. Close and clean up old shuffleMergeId files,
+          // happens in the non-deterministic stage retries
+          logger.info("Creating a new attempt for shuffle blocks push request for shuffle {} with"
+            + " shuffleMergeId {} for application {}_{} since it is higher than the latest"
+              + " shuffleMergeId {} already seen", shuffleId, shuffleMergeId, appShuffleInfo.appId,
+              appShuffleInfo.attemptId, latestShuffleMergeId);
+          if (latestShuffleMergeId != UNDEFINED_SHUFFLE_MERGE_ID &&
+              shuffleMergePartitionsMap.containsKey(latestShuffleMergeId)) {
+            Map<Integer, AppShufflePartitionInfo> latestShufflePartitions =
+              shuffleMergePartitionsMap.get(latestShuffleMergeId);
+            mergedShuffleCleaner.execute(() ->
+              closeAndDeletePartitionFiles(latestShufflePartitions));
+            shuffleMergePartitionsMap.put(latestShuffleMergeId, STALE_SHUFFLE_PARTITIONS);
+          }
+          // Remove older shuffleMergeIds which won't be required anymore
+          if (secondLatestShuffleMergeId != UNDEFINED_SHUFFLE_MERGE_ID) {
+            shuffleMergePartitionsMap.remove(secondLatestShuffleMergeId);
+          }
+          Map<Integer, AppShufflePartitionInfo> newShufflePartitions = new ConcurrentHashMap<>();
+          shuffleMergePartitionsMap.put(shuffleMergeId, newShufflePartitions);
+          shuffleMergePartitionsRef.set(shuffleMergePartitionsMap);
+          return shuffleMergePartitionsMap;
         }
-      });
-    if (shufflePartitions == null) {
-      return null;
+      }
+    });
+
+    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
+        shuffleMergePartitionsRef.get().get(shuffleMergeId);
+    if (null == shufflePartitions || isStaleOrTooLate(shufflePartitions)) {
+      // It only gets here when shufflePartitions is either FINALIZED_SHUFFLE_PARTITIONS or
+      // STALE_SHUFFLE_PARTITIONS or null. This happens in 2 cases:
+      // 1. Incoming block request is for an older shuffleMergeId of a shuffle (i.e already
+      // higher shuffle sequence Id blocks are being merged for this shuffle Id.
+      // 2. Shuffle for the current shuffleMergeId is already finalized.
+      return shufflePartitions == FINALIZED_SHUFFLE_PARTITIONS ?
+        FINALIZED_APP_SHUFFLE_PARTITION_INFO : STALE_APP_SHUFFLE_PARTITION_INFO;
```

Comment:
Took a look at this again...Looks like `STALE_APP_SHUFFLE_PARTITION_INFO` is an impossible case here.

Reply:
Lets say if we have `shuffle Id` = 0 and latest `shuffleMergeId` = 3 and if there is a push request for `shuffleMergeId` = 2 then the key `shuffleMergeId = 2` in the Map would be pointing to `STALE_SHUFFLE_PARTITIONS`, therefore this would return `STALE_APP_SHUFFLE_PARTITION_INFO` right?

Reply:
@venkata91  In this case, isn't we reject it and throw RuntimeException above?

Reply:
Thats a good point, currently if the key exists in `shuffleMergePartitions` we are returning it rather we can throw an exception directly there. so need to reorder the if/else conditions a bit.